### ISSUE TITLE
Telnet: Manhole module not available from twisted.conch for Python 3+

### ISSUE THREAD

**Issue by lucab0ni** [scrapy/scrapy#2155]
manhole module from twisted.conch has not been ported for python versions 3+.

Actual workaround I used:
1. Install manhole separately with pip
2. Modify scrapy.extensions.telnet.py and try to `import manhole` in case the import of twisted.conch.manhole has failed. If `import manhole` succeed, set variable `TWISTED_CONCH_AVAILABLE = True`

**Comment by pawelmhm** [comment]
There's ticket for porting manhole in Twisted issue tracker, latest update says there is something in progress. Worth tracking this related ticket to see how it goes https://twistedmatrix.com/trac/ticket/8327

**Comment by redapple** [comment]
@lucab0ni , @pawelmhm , as a side note,
while at it, this telnet extension could be moved to scrapy-plugins.
what do you think? cc @kmike , @dangra , @eliasdorneles

**Comment by eliasdorneles** [comment]
@redapple would the idea be to remove the feature from Scrapy or would Scrapy depend on it?

If the idea is to remove the feature from Scrapy, +1.

If the idea is to make Scrapy depend on it, it would make only sense if the code can be made completely independent of Scrapy (to avoid a cyclic dependency case, where Scrapy depends on it and it depends on Scrapy), and I'm not sure that is possible.

**Comment by kmike** [comment]
I think someone (@daniel? @pablohoffman?) was against removing telnet from Scrapy. It makes sense to have debugging features built in. We had a similar discussion about TELNET_ENABLED option: let's say it is False by default. Spider starts misbehaving, you want to debug it, then you learn about telnet extension and want to use it, but telnet is off by default, so you can't. You can kill the spider, change the option and start the spider again, but you may loose the specific state which is necessary to debug the problem.

The advantage of having manhole separate and making Scrapy depend on it is that it will allow separate release cycles. I'm not sure how important is it.

**Comment by redapple** [comment]
I see it as an extension and don't think Scrapy would need to depend on it.
My comment is mainly motivated by currently having a non-Python3 compatible extension as built-in (although this could be a short lived situation).
As for debugging, I see it similarly to a debugger, where you'd add breakpoints after seeing problems with it.
There's definitely room for improvements in debug information (stats, more actionable memdebug, saving state when killing spiders etc.). Realtime debugging is not easy. And I wonder if the telnet console is the best option; I'm no user of the telnet console myself so I don't have a strong opinion on it.

**Comment by eliasdorneles** [comment]
@lucab0ni how well did python-manhole (manhole from pip) worked for you, using your workaround?

**Comment by kmike** [comment]
The problem with telnet vs debugger is that for regular programs usually you can run a debugger and get the same results between runs, but with spiders iterations can be very slow: it is possible you won't get the same error condition if you restart a spider, or you may have to wait for a long time (maybe even days) to get it again.

**Comment by lucab0ni** [comment]
@eliasdorneles it just worked fine. That fixed the problem.

**Comment by redapple** [comment]
@lucab0ni , could you share your patch?
I tried the following changes locally against the "master" branch:
```
$ git diff
diff --git a/scrapy/extensions/telnet.py b/scrapy/extensions/telnet.py
index d9add1d..75fea82 100644
--- a/scrapy/extensions/telnet.py
+++ b/scrapy/extensions/telnet.py
@@ -9,10 +9,17 @@ import logging
 
 from twisted.internet import protocol
 try:
-    from twisted.conch import manhole, telnet
+    from twisted.conch import manhole
+except ImportError as e:
+    print(e)
+    import manhole
+
+try:
+    from twisted.conch import telnet
     from twisted.conch.insults import insults
     TWISTED_CONCH_AVAILABLE = True
-except ImportError:
+except ImportError as e:
+    print(e)
     TWISTED_CONCH_AVAILABLE = False
 
 from scrapy.exceptions import NotConfigured
diff --git a/setup.py b/setup.py
index 92c114a..5a49e0c 100644
--- a/setup.py
+++ b/setup.py
@@ -51,5 +51,6 @@ setup(
         'parsel>=0.9.3',
         'PyDispatcher>=2.0.5',
         'service_identity',
+        'manhole',
     ],
```

and it's failing to import `twisted.conch.insults`, failing to load the telnet extension.
Am I missing or misunderstanding something?

```
$ scrapy version -v
Scrapy    : 1.2.0dev2
lxml      : 3.6.0.0
libxml2   : 2.9.3
Twisted   : 16.3.1
Python    : 3.5.1+ (default, Mar 30 2016, 22:46:26) - [GCC 5.3.1 20160330]
pyOpenSSL : 16.0.0 (OpenSSL 1.0.2g-fips  1 Mar 2016)
Platform  : Linux-4.4.0-31-generic-x86_64-with-Ubuntu-16.04-xenial
```

**Comment by nyov** [comment]
@redapple, @kmike, et al: What about moving telnet to scrapy-plugins anyway, and make it a "soft-hard" dependency? I'm still in favor of decoupling it.
It could still be added as a requirement for installation, or in case it isn't available a big User Warning could be printed at startup.
Something like "TELNET_ENABLED defaults to True, but scrapy-telnet could not be loaded. You will not be able to debug your Spider in this run, if you continue." ?

**Comment by redapple** [comment]
@lucab0ni , I'm still not able to make it work with Twisted 16.4 which has a lot of `twisted.conch` modules ported to Python 3 (but not `twisted.conch.manhole`).
Could you share your patch if you confirm you made it work with Python 3? Thx.

**Comment by rolando** [comment]
See https://github.com/twisted/twisted/pull/633

**Comment by rolando** [comment]
It works with latest dev version!

```
Connected to localhost.
>>>
>>> est()
Execution engine status

time()-engine.start_time                        : 2.949514150619507
engine.has_capacity()                           : False
len(engine.downloader.active)                   : 0
engine.scraper.is_idle()                        : False
engine.spider.name                              : extract
engine.spider_is_idle(engine.spider)            : True
engine.slot.closing                             : False
len(engine.slot.inprogress)                     : 0
len(engine.slot.scheduler.dqs or [])            : 0
len(engine.slot.scheduler.mqs)                  : 0
len(engine.scraper.slot.queue)                  : 0
len(engine.scraper.slot.active)                 : 0
engine.scraper.slot.active_size                 : 0
engine.scraper.slot.itemproc_size               : 0
engine.scraper.slot.needs_backout()             : False


>>> import sys; print(sys.version)
3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12)
[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]
>>>
```

```
$ scrapy version -v
Scrapy    : 1.2.0
lxml      : 3.6.4.0
libxml2   : 2.9.2
Twisted   : 16.6.0dev0
Python    : 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12) - [GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]
pyOpenSSL : 16.2.0 (OpenSSL 1.0.2j  26 Sep 2016)
Platform  : Darwin-15.6.0-x86_64-i386-64bit
```

**Comment by redapple** [comment]
This was fixed in Twisted 17.1.
One needs at least Scrapy 1.3.2 to be able to use it.

**Comment by mdaslamn92** [comment]
c:\users\local\programs\python\python37-32\lib\site-packages\scrapy\extensions\telnet.py", line 14, in <module>
    from twisted.conch import manhole, telnet
c:\users\aslam\appdata\local\programs\python\python37-32\lib\site-packages\twisted\conch\manhole.py", line 154
    def write(self, data, async=False):
                              ^
SyntaxError: invalid syntax
